packages/test/test-gc-sweep-tests/src/test/gcSweepTest.spec.ts

```diff
@@ -0,0 +1,355 @@
+        uploadedSummaries.push(summaryTree);
+
+        await sleep(defaultSessionExpiryDurationMs);
+
+        await loadContainer();
+        assert(mainContainer.closed === true, "Expected mainContainer to be closed!");
+        assert(uploadedSummaries.length > 0, `${uploadedSummaries.length}`);
+    });
+
+    const randomInteger = (min: number, max: number): number => {
+        return makeRandom().integer(min, max);
+    };
+
+    const loadNewContainer = async () => {
+        const newContainer = await loadContainer();
+        newContainer.on("closed", () => {
+            const index = connectedContainers.indexOf(newContainer);
+            assert(index >= 0, "Expected container to have been added to connectedContainers");
+            closedContainers.push(connectedContainers[index]);
+            connectedContainers.splice(index, 1);
+        });
+        connectedContainers.push(newContainer);
+    };
+
+    const closeRandomContainer = async () => {
+        if (connectedContainers.length <= 0) {
+            return;
+        }
+        const index = randomInteger(0, connectedContainers.length - 1);
+        return new Promise((resolve) => resolve(connectedContainers[index].close()));
+    };
+
+    const createDataStoreForContainer = async (container: IContainer, dataStoreType: string = "TestDataObject") => {
+        const defaultDataStore = await requestFluidObject<TestDataObject>(container, "default");
+        const newDataStore = await defaultDataStore.containerRuntime.createDataStore(dataStoreType);
+        const newDataObject = await requestFluidObject<TestDataObject>(newDataStore, "");
+        testNodes.set(newDataObject.id, new TestNode(newDataObject.id));
+        unrefDataStores.add(newDataObject.id);
+        return newDataObject;
+    };
+
+    const createDataStoreForRandomContainer = async () => {
+        if (connectedContainers.length <= 0) {
+            return;
+        }
+        await createDataStoreForContainer(getRandomContainer());
+    };
+
+    const getTestNode = (id: string): TestNode => {
+        const testNode = testNodes.get(id);
+        assert(testNode !== undefined, `Datastore ${id} should exist in datastores!`);
+        return testNode;
+    };
+
+    const updateReferencedDataStores = (id: string) => {
+        if (!refDataStores.has(id)) {
+            refDataStores.add(id);
+            unrefDataStores.delete(id);
+            for (const childId of getTestNode(id).children) {
+                updateReferencedDataStores(childId);
+            }
+        }
+    };
+
+    const hasReferencedParent = (id: string): boolean => {
+        for (const parentNode of getTestNode(id).parents) {
+            if (refDataStores.has(parentNode)) {
+                return true;
+            }
+        }
+        return false;
+    };
+
+    const updateUnreferencedDataStores = (id: string) => {
+        if (!unrefDataStores.has(id) && !hasReferencedParent(id)) {
```

Comment:
Why don't we want to do anything if the parent is referenced?

Reply:
Not sure I totally follow, but I haven't got to the usage of these yet.

I'm wondering if you could encapsulate all the node/reference tracking into a class. I believe the point is to "shadow" the runtime's object graph and track which dataStores (by ID) you expect to be considered unreferenced at any given time. Right?

Reply:
Like maybe instead of refDataStores and unrefDataStores, the `TestNode` class could have a flag (or tri-state or whatever you need) to express its expected reference state.

Reply:
Actually you should be able to just compute referenced/unreferenced by following parent links up to a root, right?  As it is `refDataStores` and `unrefDataStores` basically constitute a cache of that state, which strikes me as error prone and not necessary for the scale of reference graphs I'd expect to see in these tests.